### ISSUE TITLE
Avoid false positive "possibly unbound"

### ISSUE THREAD

**Issue by 108anup** [microsoft/pyright#4087]
When variables are conditionally declared, then pyright may show "possibly unbound" error, even though ideally it should be able to infer that the variable won't be unbound.

```
DEFINE_X = True
if(DEFINE_X):
  x = 3

# ...
# Code that does not touch DEFINE_X
# ...

if(DEFINE_X):
  y = x + 3 # This line is highlighted as "x possibly unbound"
  # There is enough information in the code to ensure that
  # x is not unbound at this time. Perhaps such false positive
  # error messages can be avoided?
```

**Comment by erictraut** [comment]
Pyright performs "type narrowing" of individual symbols, but it doesn't not do conditional narrowing of a symbol based on the types of other symbols. This would be infeasible computationally, and I'm not aware of any static code analyzer in any language that does this.

There are many ways to work around this in your code to accommodate a static type checker. For example:

```python
x = None
if some_condition:
    x = 3

# Do something else

if x is not None:
    y = x + 3
```
